### ISSUE TITLE
Serialization issue/regression going from preview2 to GA (deserialization causing null/empty properties)

### ISSUE THREAD

**Issue by christiansparre** [dotnet/orleans#8275]
We are trying to update from 4.0-preview2 to 7.0, but have run into what appears to be a serialization issue/regression.

It manifests itself when having an inheritance hierarchy like the following:

```csharp
[GenerateSerializer]
public record Fruit([property: Id(1)] string Name);

[GenerateSerializer]
public record Apple(string Name) : Fruit(Name);
```
Notice that the `Apple` record itself does not define properties of its own.

Now when this is used in a type like the following:

```csharp
[GenerateSerializer]
public class WithListOfObject
{
    [Id(1)]
    public List<object> Items { get; set; }

    [Id(2)]
    public Foo Bar { get; set; }
}
```
The `Bar` property is null after deserialization for some reason. The `Items` list is deserialized correctly.

Now take a type that inherits from `Fruit` like this that does have a property itself:

```csharp
[GenerateSerializer]
public record Pear([property: Id(1)] int Id, string Name) : Fruit(Name);
```
Then the `Bar` property is deserialized correctly.

This is blocking us as it happens using streaming when deserializing `AzureQueueBatchContainerV2` payloads. Here the `StreamId` property ends up empty causing exceptions in `PersistentStreamPullingAgent`

I have put an example demonstrating the issue here: https://github.com/christiansparre/orleans_preview2_vs_ga_serialization_issue with the `main` branch being 4.0-preview2, where it works as expected and a branch `orleans-7` where it does not.

A very short term fix to unblock us would be to add a private field like so:

```csharp
[Id(1)]
private int _f = 0;
```
To all the affected types, but I would rather not approach our production deadline with that in place.

**Comment by christiansparre** [comment]
I think I have identified the commit where this changed https://github.com/dotnet/orleans/pull/8089

I tested it by adding the tests from the above repo to the `Orleans.Serialization.UnitTests` project and testing the two adjacent commits and our issue does not exist in https://github.com/dotnet/orleans/commit/df82177e5ff3e0e60428c855d2ba4aa767eb05cf

**Comment by christiansparre** [comment]
Fixed in #8289

Thank you @ReubenBond for the quick action 👍
